From 7c24a5b8d406bea45f22166be8bee6749be6b028 Mon Sep 17 00:00:00 2001
From: Rong Bao <webmaster@csmantle.top>
Date: Thu, 30 Apr 2026 18:00:21 +0800
Subject: [PATCH 15/19] Bug 2023167 - Part 7: Add support for
 DwarfCFIToModule::UndefinedRule. r=#profiler-reviewers!

This adds support for undefining calculation rule for a given (address,
reg) pair.

Differential Revision: https://phabricator.services.mozilla.com/D297845
---
 tools/profiler/lul/LulDwarf.cpp           |  4 ++++
 tools/profiler/lul/LulDwarfSummariser.cpp | 18 ++++++++++++++++++
 tools/profiler/lul/LulDwarfSummariser.h   |  2 ++
 3 files changed, 24 insertions(+)

diff --git a/tools/profiler/lul/LulDwarf.cpp b/tools/profiler/lul/LulDwarf.cpp
index 20f1cfdea497c831a1febcbf5fba66746ebcc47a..0a051db11943cdb555e47eeefcb59b5dc62a8bef 100644
--- a/tools/profiler/lul/LulDwarf.cpp
+++ b/tools/profiler/lul/LulDwarf.cpp
@@ -2452,6 +2452,10 @@ const UniqueString* DwarfCFIToModule::RegisterName(int i) {
 }
 
 bool DwarfCFIToModule::UndefinedRule(uint64 address, int reg) {
+  if (registerIsTracked(static_cast<DW_REG_NUMBER>(reg))) {
+    summ_->UndefineRule(address, reg);
+    return true;
+  }
   reporter_->UndefinedNotSupported(entry_offset_, RegisterName(reg));
   // Treat this as a non-fatal error.
   return true;
diff --git a/tools/profiler/lul/LulDwarfSummariser.cpp b/tools/profiler/lul/LulDwarfSummariser.cpp
index 1bc827877e3bfa788395b99492a396686d725d8e..265ffe30fec2e1d40521e26221fb3d65ee7e8e65 100644
--- a/tools/profiler/lul/LulDwarfSummariser.cpp
+++ b/tools/profiler/lul/LulDwarfSummariser.cpp
@@ -688,6 +688,24 @@ cant_summarise:
   }
 }
 
+void Summariser::UndefineRule(uintptr_t aAddress, int aReg) {
+  aAddress += mTextBias;
+  if (mCurrAddr < aAddress) {
+    // Flush the existing summary first.
+    mSecMap->AddRuleSet(&mCurrRules, mCurrAddr, aAddress - mCurrAddr);
+    if (DEBUG_SUMMARISER) {
+      mLog("LUL  ");
+      mCurrRules.Print(mCurrAddr, aAddress - mCurrAddr, mLog);
+      mLog("\n");
+    }
+    mCurrAddr = aAddress;
+  }
+  LExpr* const expr = mCurrRules.ExprForRegno(static_cast<DW_REG_NUMBER>(aReg));
+  if (expr) {
+    *expr = LExpr();
+  }
+}
+
 uint32_t Summariser::AddPfxInstr(PfxInstr pfxi) {
   return mSecMap->AddPfxInstr(pfxi);
 }
diff --git a/tools/profiler/lul/LulDwarfSummariser.h b/tools/profiler/lul/LulDwarfSummariser.h
index ff82217e9e2b9124fb15446712b2567d0afef2b2..4545e3c159feb75be148048af46393c676bbdbc0 100644
--- a/tools/profiler/lul/LulDwarfSummariser.h
+++ b/tools/profiler/lul/LulDwarfSummariser.h
@@ -24,6 +24,8 @@ class Summariser {
   // rule permanently.
   virtual void Rule(uintptr_t aAddress, int aNewReg, LExprHow how,
                     int16_t oldReg, int64_t offset);
+  // Undefine the current value recovery rule for |aReg| at |aAddress|.
+  virtual void UndefineRule(uintptr_t aAddress, int aReg);
 
   virtual uint32_t AddPfxInstr(PfxInstr pfxi);
 
-- 
2.55.0

